src/backend/migrations/cassandra/20221123_Issuer_NFTokens.cpp

```diff
@@ -0,0 +1,154 @@
+     */
+    std::stringstream query;
+    query << "SELECT hash FROM " << backend.tablePrefix()
+          << "nf_token_transactions";
+    CassStatement* nftTxQuery = cass_statement_new(query.str().c_str(), 0);
+    cass_statement_set_paging_size(nftTxQuery, 1000);
+    cass_bool_t morePages = cass_true;
+
+    // For all NFT txs, paginated in groups of 1000...
+    while (morePages)
+    {
+        std::vector<NFTsData> toWrite;
+        CassFuture* fut =
+            cass_session_execute(backend.cautionGetSession(), nftTxQuery);
+        CassResult const* result = cass_future_get_result(fut);
+        if (result == nullptr)
+            throw std::runtime_error(
+                "Unexpected empty result from nf_token_transactions");
+
+        // For each tx in page...
+        CassIterator* txPageIterator = cass_iterator_from_result(result);
+        while (cass_iterator_next(txPageIterator))
+        {
+            cass_byte_t const* buf;
+            std::size_t bufSize;
+
+            CassError rc = cass_value_get_bytes(
+                cass_row_get_column(cass_iterator_get_row(txPageIterator), 0),
+                &buf,
+                &bufSize);
+            if (rc != CASS_OK)
+                throw std::runtime_error(
+                    "Could not retrieve hash from nf_token_transactions");
+
+            auto const txHash = ripple::uint256::fromVoid(buf);
+            auto const tx = backend.fetchTransaction(txHash, yield);
+            if (!tx)
+            {
+                std::stringstream ss;
+                ss << "Could not fetch tx with hash "
+                   << ripple::to_string(txHash);
+                throw std::runtime_error(ss.str());
+            }
+            ripple::STTx const sttx{ripple::SerialIter{
+                tx->transaction.data(), tx->transaction.size()}};
+            if (sttx.getTxnType() != ripple::TxType::ttNFTOKEN_MINT)
+                continue;
+
+            ripple::TxMeta const txMeta{
+                sttx.getTransactionID(), tx->ledgerSequence, tx->metadata};
+            toWrite.push_back(
+                std::get<1>(getNFTDataFromTx(txMeta, sttx)).value());
+        }
+
+        // write what we have
+        backend.writeNFTs(std::move(toWrite));
+
+        morePages = cass_result_has_more_pages(result);
+        if (morePages)
+            cass_statement_set_paging_state(nftTxQuery, result);
+        cass_future_free(fut);
+        cass_result_free(result);
+        cass_iterator_free(txPageIterator);
+    }
+
+    cass_statement_free(nftTxQuery);
+
+    /*
+     * Step 2 - Pull every object from our initial ledger and load all NFTs
+     * found in any NFTokenPage object. Prior to this migration, we were not
+     * pulling out NFTs from the initial ledger, so all these NFTs would be
+     * missed. This will also record the URI of any NFTs minted prior to the
+     * start sequence.
+     */
+    // TODO need to test on a fresh DB. does it short circuit?
```

Comment:
Did you test yet and what does it mean exactly?

Reply:
have not tested that condition yet. there are a few conditions i need to test, but it takes a really fucking long time to seed the old data using `develop` before it's ready for a test run of the migration. i'm in the process of seeding all  the NFT data back to october to test this. i think we should continue with the code review and other parts while i'm making progress there